File: StandardWorks/SearchView.swift
import SwiftUI

/// SEARCH, NATIVE, ALL SIX VOLUMES.
///
/// Results come from the site's cross-volume index (SearchIndex) as you type,
/// grouped by volume in the canon's order; a tap opens the verse in the Read
/// tab through the same deep link the home page's search uses.
struct SearchView: View {
    @EnvironmentObject var shell: WebShell
    @State private var hits: [SearchIndex.Hit] = []
    @State private var searched = ""
    @State private var indexReady = false
    /// A volume to search within, or "" for the whole canon.
    @State private var scope = ""
    private var query: Binding<String> { $shell.searchQuery }

    private var grouped: [(String, [SearchIndex.Hit])] {
        var order: [String] = [], byVol: [String: [SearchIndex.Hit]] = [:]
        for h in hits {
            if byVol[h.volumeName] == nil { order.append(h.volumeName) }
            byVol[h.volumeName, default: []].append(h)
        }
        return order.map { ($0, byVol[$0] ?? []) }
    }

    var body: some View {
        NavigationStack {
            List {
                Group {
                    if !indexReady && !shell.searchQuery.isEmpty {
                        HStack(spacing: 10) { ProgressView(); Text("Preparing the index\u{2026}").foregroundStyle(shell.ink2) }
                    } else if hits.isEmpty && searched.count >= 2 {
                        Text("No verses match \u{201c}\(searched)\u{201d}.").foregroundStyle(shell.ink2)
                    } else if hits.isEmpty && shell.searchQuery.isEmpty {
                        Text("Hebrew, with or without vowels, or English \u{2014} every volume at once.")
                            .font(ShellTheme.text(.footnote)).foregroundStyle(shell.ink2)
                    }
                }
                .shellRow(shell)
                ForEach(grouped, id: \.0) { name, list in
                    Section {
                        ForEach(list) { hit in
                            Button {
                                // the whole list goes with it: the find bar walks it verse by verse
                                if let i = hits.firstIndex(where: { $0.id == hit.id }) {
                                    shell.startFind(query: searched, hits: hits, at: i)
                                } else {
                                    shell.open(path: hit.path)
                                }
                            } label: {
                                VStack(alignment: .leading, spacing: 4) {
                                    Text(hit.row.ref).font(ShellTheme.text(.footnote, weight: .semibold)).foregroundStyle(shell.here)
                                    if SearchIndex.hasHebrew(searched) {
                                        Text(hit.snippet).font(ShellTheme.hebrew(18))
                                            .frame(maxWidth: .infinity, alignment: .trailing)
                                            .environment(\.layoutDirection, .rightToLeft)
                                    } else {
                                        Text(hit.snippet).font(ShellTheme.text(.subheadline)).foregroundStyle(shell.ink2).lineLimit(3)
                                    }
                                }
                            }
                            .foregroundStyle(shell.ink)
                        }
                    } header: {
                        Text("\(name) \u{00B7} \(list.count)\(list.count >= 200 ? "+" : "")").foregroundStyle(shell.ink2)
                    }
                    .shellRow(shell)
                }
            }
            .listStyle(.insetGrouped)
            .shellPage(clearOfRow: true)
            .shellBar("Search")
            // The field under the title, always: from iOS 26 the default puts
            // it at the bottom of the screen, under the floating row, where it
            // could not be seen or reached (user, 2026-09-24, on iOS 27).
            .searchable(text: query, isPresented: $shell.searchPresented,
                        placement: .navigationBarDrawer(displayMode: .always), prompt: "Hebrew or English")
            .searchScopes($scope, activation: .onSearchPresentation) {
                Text("All").tag("")
                ForEach(shell.volumes) { v in Text(v.short).tag(v.key) }
            }
            .onChange(of: scope) { _, _ in run(shell.searchQuery) }
            .onSubmit(of: .search) { run(shell.searchQuery) }
            .task(id: shell.searchQuery) {
                // As you type, after a beat; the index is the whole canon.
                try? await Task.sleep(nanoseconds: 250_000_000)
                if !Task.isCancelled { run(shell.searchQuery) }
            }
            .onReceive(shell.searchIndex.$loaded) { ready in
                indexReady = ready
                if ready { run(shell.searchQuery) }
            }
        }
    }

    private func run(_ q: String) {
        let trimmed = q.trimmingCharacters(in: .whitespaces)
        searched = trimmed
        let all = trimmed.count >= 2 ? shell.searchIndex.find(trimmed) : []
        hits = scope.isEmpty ? all : all.filter { $0.row.volume == scope }
    }
}

File: android/app/src/main/java/com/sefermormon/standardworks/SearchView.kt
package com.sefermormon.standardworks

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalLayoutDirection
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.runtime.CompositionLocalProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext

/**
 * SEARCH, NATIVE, ALL SIX VOLUMES. Results come from the site's cross-volume
 * index (SearchIndex) as you type, grouped by volume in the canon's order,
 * with a scope for each volume; a tap opens the verse in the Read tab.
 */
@Composable
fun SearchView(shell: WebShell) {
    val p = shell.palette
    var scope by remember { mutableStateOf("") }
    var hits by remember { mutableStateOf(listOf<SearchIndex.Hit>()) }
    var searched by remember { mutableStateOf("") }
    val ready = shell.searchIndex.loaded

    LaunchedEffect(shell.searchQuery, scope, ready) {
        delay(250)
        val q = shell.searchQuery.trim()
        searched = q
        val all = if (q.length >= 2 && ready) withContext(Dispatchers.Default) { shell.searchIndex.find(q) } else emptyList()
        hits = if (scope.isEmpty()) all else all.filter { it.row.volume == scope }
    }

    val grouped = remember(hits) {
        val order = ArrayList<String>(); val by = LinkedHashMap<String, MutableList<SearchIndex.Hit>>()
        for (h in hits) { if (by[h.volumeName] == null) { order.add(h.volumeName); by[h.volumeName] = ArrayList() }; by[h.volumeName]!!.add(h) }
        order.map { it to by[it]!! }
    }

    ShellPage(shell, title = "Search", barContent = {
        // The field and the scope row live in the navy bar, as they do on the iPhone.
        Column(Modifier.padding(horizontal = 16.dp, vertical = 8.dp)) {
            Row(Modifier.fillMaxWidth().height(40.dp).clip(RoundedCornerShape(10.dp)).background(p.onChrome.copy(alpha = 0.14f)).padding(horizontal = 10.dp),
                verticalAlignment = Alignment.CenterVertically) {
                Icon(Icons.Outlined.Search, null, tint = p.onChrome.copy(alpha = 0.7f), modifier = Modifier.size(20.dp))
                Spacer(Modifier.width(8.dp))
                Box(Modifier.weight(1f)) {
                    if (shell.searchQuery.isEmpty()) Text("Hebrew or English", color = p.onChrome.copy(alpha = 0.6f), fontSize = 17.sp)
                    BasicTextField(
                        value = shell.searchQuery, onValueChange = { shell.searchQuery = it },
                        singleLine = true,
                        textStyle = TextStyle(color = p.onChrome, fontSize = 17.sp),
                        cursorBrush = SolidColor(p.hereChrome),
                        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                        keyboardActions = KeyboardActions(onSearch = {}),
                        modifier = Modifier.fillMaxWidth()
                    )
                }
                if (shell.searchQuery.isNotEmpty()) {
                    Icon(Icons.Outlined.Close, "Clear", tint = p.onChrome.copy(alpha = 0.8f),
                        modifier = Modifier.size(20.dp).clip(CircleShape).clickable { shell.searchQuery = "" })
                }
            }
            Spacer(Modifier.height(8.dp))
            Row(Modifier.horizontalScroll(rememberScrollState())) {
                ScopeChip(p, "All", scope == "") { scope = "" }
                for (v in shell.volumes) { Spacer(Modifier.width(6.dp)); ScopeChip(p, v.short, scope == v.key) { scope = v.key } }
            }
        }
    }) {
        LazyColumn(contentPadding = PaddingValues(16.dp)) {
            val q = shell.searchQuery
            if (!ready && q.isNotEmpty()) item { Text("Preparing the index…", color = p.ink2, modifier = Modifier.padding(16.dp)) }
            else if (hits.isEmpty() && searched.length >= 2) item { Text("No verses match “$searched”.", color = p.ink2, modifier = Modifier.padding(16.dp)) }
            else if (hits.isEmpty() && q.isEmpty()) item { Text("Hebrew, with or without vowels, or English — every volume at once.", color = p.ink2, fontSize = 13.sp, modifier = Modifier.padding(16.dp)) }
            grouped.forEachIndexed { gi, (name, list) ->
                item {
                    if (gi > 0) Spacer(Modifier.height(20.dp))
                    SectionHeader(p, "$name · ${list.size}${if (list.size >= 200) "+" else ""}")
                    ShellCard(p) {
                        list.forEachIndexed { i, hit ->
                            if (i > 0) Rule(p)
                            ShellRow(p, onClick = { shell.open(hit.path) }) {
                                Column(Modifier.weight(1f)) {
                                    Text(hit.row.ref, color = p.here, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
                                    Spacer(Modifier.height(4.dp))
                                    if (SearchIndex.hasHebrew(searched)) {
                                        CompositionLocalProvider(LocalLayoutDirection provides LayoutDirection.Rtl) {
                                            Text(hit.snippet, color = p.ink, fontSize = 18.sp, fontFamily = ShellTheme.hebrew, textAlign = TextAlign.Start, modifier = Modifier.fillMaxWidth())
                                        }
                                    } else {
                                        Text(hit.snippet, color = p.ink2, fontSize = 15.sp, maxLines = 3, overflow = TextOverflow.Ellipsis)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun ScopeChip(p: AppShell.Palette, label: String, selected: Boolean, onClick: () -> Unit) {
    Box(
        Modifier.heightIn(min = 32.dp).clip(RoundedCornerShape(8.dp))
            .background(if (selected) p.onChrome.copy(alpha = 0.22f) else p.onChrome.copy(alpha = 0.08f))
            .border(1.dp, p.onChrome.copy(alpha = if (selected) 0.35f else 0.12f), RoundedCornerShape(8.dp))
            .clickable(onClick = onClick).padding(horizontal = 12.dp, vertical = 6.dp),
        contentAlignment = Alignment.Center
    ) { Text(label, color = p.onChrome, fontSize = 14.sp, fontWeight = if (selected) FontWeight.SemiBold else FontWeight.Normal) }
}
